kotlin/src/main/kotlin/net/jqwik/kotlin/api/CombinatorDsl.kt

```diff
@@ -0,0 +1,97 @@
+package net.jqwik.kotlin.api
+
+import net.jqwik.api.Arbitrary
+import org.apiguardian.api.API
+import java.util.Optional
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KProperty
+
+
+/**
+ * Combine arbitraries using the combinator DSL:
+ *
+ * ```kotlin
+ * combine {
+ *     val first by Arbitraries.strings()
+ *     val second by Arbitraries.strings()
+ *
+ *     createAs {
+ *         "first: $first, second: $second"
+ *     }
+ * }
+ * ```
+ *
+ * @return new Arbitrary instance
+ */
+@API(status = API.Status.EXPERIMENTAL, since = "1.8.0")
+fun <R> combine(
+    combinator: CombinatorScope.() -> Combined<R>
+): Arbitrary<R> {
+    val combined = CombinatorScope().combinator()
+
+    @Suppress("UNCHECKED_CAST")
+    return combine(combined.arbitraries as List<Arbitrary<Any?>>) { values ->
+        try {
+            for ((property, value) in combined.properties zip values) {
+                property.bind(value)
+            }
+
+            CombinatorCreateScope().(combined.combinator)()
+        } finally {
+            for (it in combined.properties) {
+                it.unbind()
+            }
+        }
+    }
+}
+
+@API(status = API.Status.EXPERIMENTAL, since = "1.8.0")
+class ArbitraryProperty<T> internal constructor() : ReadOnlyProperty<Nothing?, T> {
+    private val currentValue = ThreadLocal<Optional<T & Any>>()
```

Comment:
Could you clarify why `ThreadLocal` is needed here? Why isn't a regular reference enough?

Reply:
I think I understand now, however, it would be great to put it in the comment

Reply:
I was assuming here that there could be a feature like parallel execution (now or in the future) that calls the same Arbitrary multiple times in parallel. If that's not the case, then `ThreadLocal` is not necessary of course.

The issue here is that the block passed to `combine` is only called once, but the block passed to `createAs` is called every time the arbitrary has to provide a value. Since the `ArbitraryProperty` is instantiated in `combine`, the instance is shared across all invocations.

This is the hacky part I was talking about 😄

Reply:
> it would be great to put it in the comment

Completely missed this comment -  my internet is being kind of flaky right now, so I'm just not receiving comments sometimes 😄

I added some documentation to `ValueBindings` (where the `ThreadLocal` is locade now) 🙂